client/zeth/core/mimc.py

```diff
@@ -2,53 +2,106 @@
-    Python implementation of MiMC7 algorithm used in the mixer contract
+    Base class of MiMC implmentations.
     """
-
     def __init__(
             self,
-            seed: str = MIMC_MT_SEED,
-            prime: int = ZETH_PRIME):
+            seed_str: str,
+            prime: int,
+            num_rounds: int):
+        self.seed = _keccak_256(_str_to_bytes(seed_str))
         self.prime = prime
-        self.seed = seed
-
-    def mimc_round(self, message: int, key: int, rc: int) -> int:
-        xored = (message + key + rc) % self.prime
-        return xored ** 7 % self.prime
+        self.num_rounds = num_rounds
 
-    def mimc_encrypt(
+    def encrypt(
             self,
             message: int,
-            ek: int,
-            seed: Optional[str] = None,
-            rounds: int = 91) -> int:
-        seed = seed or self.seed
-        res = message % self.prime
+            ek: int) -> int:
+        result = message % self.prime
         key = ek % self.prime
+        round_constant: int = self.seed
 
-        # In the paper the first round constant is set to 0
-        res = self.mimc_round(res, key, 0)
+        # The rc in round 0 is 0 (see [AGRRT16])
+        result = self.mimc_round(result, key, 0)
 
-        round_constant: int = _keccak_256(_str_to_bytes(seed))
+        for _ in range(self.num_rounds - 1):
+            round_constant = _update_round_constant(round_constant)
+            result = self.mimc_round(result, key, round_constant)
 
-        for _ in range(rounds - 1):
-            round_constant = _keccak_256(_int_to_bytes32(round_constant))
-            res = self.mimc_round(res, key, round_constant)
+        # Add key to the final result (see [AGRRT16])
+        return (result + key) % self.prime
 
-        return (res + key) % self.prime
-
-    def mimc_mp(self, x: int, y: int) -> int:
+    def hash(self, x: int, y: int) -> int:
+        """
+        Apply Miyaguchi-Preneel to the output of the encrypt function.
+        """
         x = x % self.prime
         y = y % self.prime
-        return (self.mimc_encrypt(x, y, self.seed) + x + y) % self.prime
+        return (self.encrypt(x, y) + x + y) % self.prime
+
+    @abstractmethod
+    def mimc_round(self, message: int, key: int, rc: int) -> int:
+        pass
+
+
+class MiMC7(MiMCBase):
+    """
+    MiMC specialized for Fr in ALT-BN128, in which the exponent is 7 and 91
+    rounds are used.
+    """
+    def __init__(self, seed_str: str = MIMC_MT_SEED):
+        MiMCBase.__init__(
+            self,
+            seed_str,
+            21888242871839275222246405745257275088548364400416034343698204186575808495617,  # noqa
+            # pylint: disable=line-too-long
+            91)
+
+    def mimc_round(self, message: int, key: int, rc: int) -> int:
+        xored = (message + key + rc) % self.prime
+        return xored ** 7 % self.prime
+
+
+class MiMC31(MiMCBase):
+    """
+    MiMC implementation using exponent of 11 and 51 rounds. Note that this is
```

Comment:
Wrong exponent (`ceil(log(rBLS,2)/log(31,2)) = 51`). May be a residual from the switch away from MiMC11

`MiMC implementation using exponent of 11 and 51 rounds` -> `MiMC implementation using exponent of 11 and 51 rounds`

Reply:
fixed